doc/UPGRADE.md

```diff
@@ -0,0 +1,428 @@
+- **Action**: Upgrade Node.js before upgrading http-server
+
+The specific patch version (16.20.2) is unusual. Verify this requirement hasn't changed.
+
+### 2. IPv6 Default Binding - SECURITY CRITICAL
+
+**This is the most dangerous change in this release.**
+
+- **Old**: Default host `0.0.0.0` (IPv4 only)
+- **New**: Default host `::` (dual-stack IPv4+IPv6)
+
+**DANGER**: Many firewalls are configured for IPv4 only. Enabling IPv6 by default means your server may be exposed to the internet BYPASSING YOUR FIREWALL.
+
+**Before upgrading**:
+1. Check if IPv6 is enabled: `ip -6 addr show` (Linux) or `ifconfig` (macOS)
+2. Verify your firewall covers IPv6: `ip6tables -L -n` (Linux)
+3. Understand that internal/dev servers may suddenly be internet-accessible
+4. Test from an external IPv6 network if possible
+
+**To maintain IPv4-only behavior**:
+```bash
+http-server -a 0.0.0.0
+```
+
+Update all deployment scripts, systemd services, and Docker configs to include this flag if you're not ready for dual-stack.
+
+### 3. Development Dependencies (Contributors Only)
+- **Old**: TAP v14
+- **New**: TAP v21
+- **Impact**: Test framework completely changed
+
+## Security Warnings
+
+### Header Injection Vulnerability (New -H Flag)
+
+**SEVERITY: CRITICAL**
+
+The new `-H/--header` option has NO CRLF validation. This allows HTTP response splitting attacks if you pass untrusted input.
+
+**NEVER do this**:
+```bash
+http-server -H "$USER_INPUT"
+```
+
+**Safe usage**:
+```bash
+# Only use with hardcoded values
+http-server -H "X-Custom: MyValue" -H "Cache-Control: max-age=3600"
+```
+
+If you need dynamic headers, validate rigorously:
+- No `\r` or `\n` characters (including URL-encoded %0D, %0A)
+- No control characters
+- Better: use a reverse proxy to set headers instead
+
+**This should be fixed in http-server itself.** Users shouldn't be responsible for security validation.
+
+### WebSocket Proxy - Don't Use in Production
+
+The new WebSocket support in proxy mode (`-P` flag) introduces attack vectors:
+- No rate limiting for WebSocket connections
+- Long-lived connections can exhaust resources
+- Authentication bypass risks
+
+**Recommendation**: Don't use http-server for production WebSocket workloads. Use nginx, HAProxy, or a dedicated WebSocket proxy. http-server is a simple file server, not a production WebSocket handler.
+
+### CORS Wildcard Warning
+
+`--cors` sets `Access-Control-Allow-Origin: *`, allowing ANY website to make JavaScript requests to your server and read responses.
+
+**DANGER**:
+- Any malicious website can access your API
+- Credentials in cookies will be sent
+- Private data becomes accessible to any origin
+- No CSRF protection
+
+**Only use --cors for**:
+- Public CDN resources (fonts, images, libraries)
+- Open APIs with no authentication
+- Development/testing (never production)
+
+**Never use --cors with**:
+- APIs with authentication
+- Any endpoint serving user-specific data
+- Internal resources
```

Comment:
As an upgrade guide, should this be here? This won't cause a regression for anyone who isn't already using the flag. Maybe this belongs in the documentation for this flag.